core/src/main/scala/kafka/log/Log.scala

```diff
@@ -407,7 +407,8 @@ class Log(@volatile var dir: File,
 
         // maybe roll the log if this segment is full
         val segment = maybeRoll(messagesSize = validMessages.sizeInBytes,
-                                maxTimestampInMessages = appendInfo.maxTimestamp)
+                                maxTimestampInMessages = appendInfo.maxTimestamp,
+                                minOffsetInMessages = appendInfo.firstOffset)
```

Comment:
It seems that we should be passing in appendInfo.lastOffset instead of firstOffset. The reason is that when appending to the time index, we use the offset of the message with the largest timestamp. That offset could be as large as the last offset.